core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,332 @@
+        let pos = 2u64.pow(depth) + index;
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
+
+        let old_hashes = self.compute_hashes(&path, depth, index, false);
+
+        // Fill new_hashes vector -----------------------------------------------------------------
+        let pos = 2u64.pow(depth) + index;
+
+        if is_even(pos) {
+            self.paths.get_mut(&(pos / 2)).unwrap()[0] = value;
+        } else {
+            self.paths.get_mut(&(pos / 2)).unwrap()[1] = value;
+        }
+
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
+
+        let new_hashes = self.compute_hashes(&path, depth, index, true);
+
+        // update paths ---------------------------------------------------------------------------
+        for (_, path) in self.paths.iter_mut() {
+            for i in (0..old_hashes.len()).rev() {
+                if path[i + 2] == old_hashes[i] {
+                    path[i + 2] = new_hashes[i];
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn compute_hashes(
+        &mut self,
+        path: &[Word],
+        depth: u32,
+        index: u64,
+        update_root: bool,
+    ) -> Vec<Word> {
+        let mut pos = 2u64.pow(depth) + index;
+        let mut path_iter = path.iter();
+
+        let mut computed_hashes = Vec::<Word>::new();
+
+        let mut path_hash;
+        let mut comp_hash;
+
+        if is_even(pos) {
+            comp_hash = *path_iter.next().unwrap();
+            path_hash = *path_iter.next().unwrap();
+        } else {
+            path_hash = *path_iter.next().unwrap();
+            comp_hash = *path_iter.next().unwrap();
+        }
+
+        for _ in 0..depth - 2 {
+            comp_hash = calculate_parent_hash(comp_hash, pos, path_hash);
+            pos /= 2;
+            computed_hashes.push(comp_hash);
+            path_hash = *path_iter.next().unwrap();
+        }
+
+        if update_root {
+            self.root = calculate_parent_hash(comp_hash, pos, path_hash);
+        }
+
+        computed_hashes
+    }
+}
+
+// HELPER FUNCTIONS
+// --------------------------------------------------------------------------------------------
+
+fn is_even(pos: u64) -> bool {
+    pos & 1 == 0
+}
+
+// pos -- position of the cur hash
+pub fn calculate_parent_hash(cur: Word, pos: u64, hash_2: Word) -> Word {
+    if is_even(pos) {
+        hasher::merge(&[cur.into(), hash_2.into()]).into()
+    } else {
+        hasher::merge(&[hash_2.into(), cur.into()]).into()
+    }
+}
+
+fn compute_root(path: &[Word], depth: u32, index: u64) -> Word {
+    let mut pos = 2u64.pow(depth) + index;
+    let mut path_iter = path.iter();
+
+    let mut path_hash;
+    let mut comp_hash;
+
+    if is_even(pos) {
+        comp_hash = *path_iter.next().unwrap();
+        path_hash = *path_iter.next().unwrap();
+    } else {
+        path_hash = *path_iter.next().unwrap();
+        comp_hash = *path_iter.next().unwrap();
+    }
```

Comment:
You might be able to simplify this. Since you know that the even node is first in the path, and the odd one is next, you should be able to compute the first `path_hash` as:
```Rust
let mut path_hash = hasher::merge(&[path[0].into(), path[1].into()]).into();
```